keps/sig-api-machinery/0006-apply.md

```diff
@@ -443,6 +445,101 @@ y,Initialized,PodScheduled,Ready,/var/run/secrets/kubernetes.io/serviceaccount]}
 | 400 in base64 | \~ | \~ | \~ |
 | 300 in raw bytes | \~ | \~ | \~ |
 
+##### Generating String Tables
+
+A traversal across the Kubernetes type definitions in the OpenAPI spec will be
+used to gather all field names and enum values, which can be used as a starting
+point for the string table.
+
+The strings found in type definitions will be supplemented with other strings
+found to be frequently used in fieldsets. Commonly used string values and
+unstructured map keys are good candidates. For example,
+`/var/run/secrets/kubernetes.io/serviceaccount` was found by direct inspection
+as being a great string table entry; it is quite long, and it is frequently
+used. We expect to find more string table entries like this via tools and direct
+inspection, and will keep track of them in a separate list that will then be
+merged together with the strings from type definitions when generating string
+tables.
+
+A couple considerations when generating the string table:
+
+- Not all strings are worth adding to the string table. If the `!<base64>`
+  string is longer than the string it would replace, then it clearly makes no
+  sense to add it to the dictionary.
+- Since the base64 strings of low ints are shorter, they should be used for
+  more frequently used strings in fieldsets.
+
+Fine tuning the string table based on what strings are most frequent is
+something we can iterate on. If the initial string table size is small, ordering
+by frequency might not matter that much, but if/when the table gets larger,
+ensuring that frequently used strings are assigned low numbers in the table will
+become increasingly valuable.
+
+There are plenty of potential future optimizations, like gathering fieldset data
+from real-world Kubernetes clusters at runtime, and using it as training data to
+improve string tables.
+
+##### Versioning String Tables
+
+Each string table has a version, starting at 1:
+
+```
+String table:
+{<versionNumber>:[...
+```
+
+Fieldsets serialized using string tables contain the version they were serialized with:
+
+```
+Field set:
+[<versionNumber>, ...
+```
+
+Since a fieldset can only be deserialized using the versioned string table
+it was serialized with, each Kubernetes version must be able to deserialize
+fieldsets that were serialized by:
+
+- All of the prior Kubernetes minor versions.
+- The next Kubernetes minor version, so that N-1 rollback is safe
+
+We can ensure this by:
+
+- Keeping all string table versions around forever
+- Introducing a string table version to Kubernetes one minor version before
+  serializing data using that string table version
+
+Alternative considered: We could instead persist string table versions in etcd--
```

Comment:
This solution addresses the most important interaction which is indeed the apiserver being able to read the objects, but it doesn't solve the problem for clients who want to look at the fieldsets.

Reply:
Well, if we exposed the string tables to clients then this could work for that too